File: project/jimmer-core/src/test/java/org/babyfish/jimmer/model/BookStore.java
package org.babyfish.jimmer.model;

import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import org.babyfish.jimmer.Immutable;

import javax.validation.constraints.*;
import java.math.BigDecimal;
import java.util.List;

@Immutable
public interface BookStore {

    @NotBlank
    @Size(max = 50)
    @Pattern(regexp = "[^\\d]+\\S+")
    @UpperCase
    String name();

    @Null
    String website();

    List<Book> books();

    @Null
    @PositiveOrZero
    BigDecimal avgPrice();
}

File: project/jimmer-core-kotlin/src/test/kotlin/org/babyfish/jimmer/kt/model/BookStore.kt
package org.babyfish.jimmer.kt.model

import org.babyfish.jimmer.Immutable
import java.math.BigDecimal
import javax.validation.constraints.Pattern
import javax.validation.constraints.PositiveOrZero

@Immutable
interface BookStore {

    @get:Pattern(regexp = "[^\\d]+\\S+")
    val name: String

    val books: List<Book>

    @get:PositiveOrZero
    val avgPrice: BigDecimal?
}